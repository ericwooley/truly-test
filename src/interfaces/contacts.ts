export interface IContact {
  id?: string;
  name: string; 
  number: string; 
  context: string; 
}
