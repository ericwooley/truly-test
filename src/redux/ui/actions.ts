import { createAction } from "redux-actions";

export const actions = {
  displayAddContactForm: "DISPLAY_ADD_CONTACT_FORM"
};

export const actionCreators = {
  displayAddContactForm: createAction<boolean>(actions.displayAddContactForm)
};
