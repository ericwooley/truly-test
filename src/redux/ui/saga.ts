// import { call } from "redux-saga/effects";
// import { eventChannel, delay } from "redux-saga";
// import { actions, actionCreators } from "./actions";
// import selectors from "./selectors";
// export function * globalEventsSaga() {
//   yield call(console.log, "ui saga loaded");
// }
