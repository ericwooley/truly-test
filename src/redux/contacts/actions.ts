import { createAction } from "redux-actions";
import { IContact } from "../../interfaces/contacts";
export const actions = {
  updateSearchValue: "UPDATE_SEARCH_VALUE",
  loadContacts: "LOAD_CONTACTS",
  setContacts: "SET_CONTACTS",
  updateNewContact: "UPDATE_NEW_CONTACT",
  resetNewContact: "RESET_NEW_CONTACT",
  createNewContact: "CREATE_NEW_CONTACT"
};
export interface IUpdateNewContactPayload {
  key: string; value: string;
}
export const actionCreators = {
  createNewContact: createAction<IContact>(actions.createNewContact),
  updateNewContact: createAction<IUpdateNewContactPayload>(actions.updateNewContact),
  resetNewContact: createAction(actions.resetNewContact),
  setContacts: createAction <IContact[]>(actions.setContacts),
  loadContacts: createAction(actions.loadContacts),
  updateSearchValue: createAction<string>(actions.updateSearchValue)
};
