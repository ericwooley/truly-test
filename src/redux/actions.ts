import {actionCreators as contacts } from "./contacts/actions";
import {actionCreators as ui } from "./ui/actions";
export default {
  contacts,
  ui
};
