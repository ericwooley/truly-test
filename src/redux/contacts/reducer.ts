import { 
  handleActions,
  Action
} from "redux-actions";
import { actions, IUpdateNewContactPayload } from "./actions";
import { IContact } from "../../interfaces/contacts";
const initialContacts: IContact[] = [];
const newContact: IContact = {
  name: "",
  context: "",
  number: ""
};
const initialState = {
  searchValue: "",
  contacts: initialContacts,
  newContact
};
export type ContactState = typeof initialState;
const reducer = handleActions(
  {
    [actions.resetNewContact]: (state: ContactState, action: Action<undefined>): ContactState => ({
      ...state,
      newContact
    }),
    [actions.updateNewContact]: (state: ContactState, action: Action<IUpdateNewContactPayload>): ContactState => {
      if (action.payload) {
        return ({
          ...state,
          newContact: {
            ...state.newContact,
            [action.payload.key]: action.payload.value
          }
        });
      }
      return state;
      
  },
    [actions.updateSearchValue]: (state: ContactState, action: Action<string>): ContactState =>
      ({...state, searchValue: action.payload || ""}),
    [actions.setContacts]: (state: ContactState, action: Action<IContact[]>): ContactState =>
      ({...state, contacts: action.payload || []})
  },
  initialState as any // this bugs me
);

export default reducer;
