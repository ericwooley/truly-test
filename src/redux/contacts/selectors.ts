import { createSelector } from "reselect";

import { IContact } from "../../interfaces/contacts";
import { ContactState } from "./reducer";
export interface IContactSubState { contacts: ContactState; }
const searchValue = (state: IContactSubState) => state.contacts.searchValue;
const contactsUnsorted = (state: IContactSubState) => state.contacts.contacts;
const contacts = createSelector(contactsUnsorted, (contactList: IContact[]) => 
  [...contactList].sort((a, b) => {
    const aName = a.name.toLowerCase();
    const bName = b.name.toLowerCase();
    if (aName < bName) {
      return -1;
    }
    if (aName > bName) {
      return 1;
    }
    return 0;
  }));
const userEnteredContact = (state: IContactSubState) => state.contacts.newContact;
const filteredContacts = createSelector(
  searchValue, 
  contacts, 
  (value: string, contactList: IContact[]) => {
    if (!value) {
      return contactList;
    }
    value = value.toLowerCase();
    return contactList.filter(c => {
      const matchesName = c.name.toLowerCase().indexOf(value) !== -1;
      const matchesNumber = c.number.toLowerCase().indexOf(value) !== -1;
      const matchesContext = c.context.toLowerCase().indexOf(value) !== -1;
      return matchesName || matchesContext || matchesNumber;
    }); 
  }
);

const userContact = (state: IContactSubState) => state.contacts.newContact;
const userContactErrors = createSelector(userContact, (contact): IContact => ({
  name: contact.name.length ? "" : "name is required",
  context: contact.context.length ? "" : "context is required",
  number:  contact.number.length ? "" : "phone number is required",
}));
export default {
  userContactErrors,
  userEnteredContact,
  contacts,
  filteredContacts,
  searchValue
};
