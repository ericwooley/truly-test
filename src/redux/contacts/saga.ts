import { call, take, all, put, select } from "redux-saga/effects";
import { getContacts } from "../../services/getContacts";
import { 
  Action
} from "redux-actions";
// import { eventChannel, delay } from "redux-saga";
import { 
  actions, 
  actionCreators 
} from "./actions";
import allActionCreatorsFrom from "../actions";
import { IContact } from "../../interfaces/contacts";
import selectors from "./selectors";
import { createContact } from "../../services/createContact";
export function * contacts() {
  yield all([
    watchForLoadContacts(),
    watchCreateContact()
  ]);
}
export function * watchForLoadContacts () {
  while (true) {
    yield take(actions.loadContacts);
    yield loadContacts();
  }
}

export function * watchCreateContact (): any {
  while (true) {
    const action: Action<IContact> = yield take(actions.createNewContact);
    if (action.payload) {
      const contactsFromState: IContact[] = yield select(selectors.contacts);
      yield createContactSaga(action.payload, contactsFromState);
    }
  }
}

export function * createContactSaga (contact: IContact, currentContactList: IContact[]): any {
  try {
    yield put(actionCreators.setContacts([contact, ...currentContactList]));
    const returnedContact = yield call(createContact, contact);
    yield put(actionCreators.setContacts([returnedContact, ...currentContactList]));
    yield put(actionCreators.resetNewContact());
    yield put(allActionCreatorsFrom.ui.displayAddContactForm(false));
  } catch (e) {
    const retry = yield call(confirm, "There was an error creating your new contact, retry?");
    if (retry) {
      yield put(actionCreators.setContacts(currentContactList));
      yield createContactSaga(contact, currentContactList);
    } else {
      // reload the contacts because we don't know where in the process the error occured
      yield loadContacts();
    }
  }
}

export function * loadContacts (): any {
  try {
    // TODO: Validate contact list
    const contactsFromApi: IContact[] = yield call(getContacts);
    yield put(actionCreators.setContacts(contactsFromApi));
  } catch (e) {
    const retry = yield call(confirm, "There was an error loading contacts, retry?");
    if (retry) {
      yield loadContacts();
    }
  }
}
