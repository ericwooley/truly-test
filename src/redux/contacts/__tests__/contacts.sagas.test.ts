import { take, call, put } from "redux-saga/effects";
import { actions, actionCreators } from "../actions";
import { watchForLoadContacts, loadContacts, createContactSaga } from "../saga";
import { getContacts } from "../../../services/getContacts";
import { IContact } from "../../../interfaces/contacts";
import { createContact } from "../../../services/createContact";
const {contacts: contactsTestData} = require("../../../../db.testdata.json");
describe("contacts sagas", () => {
  describe("watchForLoadContacts", () => {
    const iterator = watchForLoadContacts();
    let loadContactsIterator: Generator;
    it("should wait for the load contacts action", () => {
      expect(iterator.next().value).toEqual(take(actions.loadContacts));
    });
    it("should return the load contacts generator", () => {
      loadContactsIterator = iterator.next().value;
      expect(typeof loadContactsIterator.next).toBe("function");
    });
  });
  describe("loadContacts", () => { 
    describe("successful call", () => {
      let loadContactsIterator = loadContacts();
      it("should get the contacts", () => {
        expect(loadContactsIterator.next().value).toEqual(call(getContacts));
      });
      it("should get the the contacts", () => {
        expect(loadContactsIterator.next(contactsTestData).value)
          .toEqual(put(actionCreators.setContacts(contactsTestData)));
      });
    });
    describe("unsuccessful call", () => {
      it("should confirm to retry", () => {
        let loadContactsIterator: Generator = loadContacts();
        if (loadContactsIterator.throw) {
          loadContactsIterator.next();
          expect(loadContactsIterator.throw(new Error("did not work")).value)
            .toEqual(call(confirm, "There was an error loading contacts, retry?"));
          expect(typeof loadContactsIterator.next(true).value.next).toEqual("function");
        } else {
          throw new Error("something went wrong");
        }
      });
      it("should stop when the user does not retry", () => {
        let loadContactsIterator: Generator = loadContacts();
        if (loadContactsIterator.throw) {
          loadContactsIterator.next();
          expect(loadContactsIterator.throw(new Error("did not work")).value)
            .toEqual(call(confirm, "There was an error loading contacts, retry?"));
          expect(loadContactsIterator.next(false).done).toEqual(true);
        } else {
          throw new Error("something went wrong");
        }
      });
    });
  });
  describe("create contact", () => { 
    const demoContact: IContact = {
      name: "steve",
      number: "1234567890",
      context: "work"
    };
    const uuid = "e02604b6-5e1d-402b-ae36-a966899fdcd9";
    // using snapshot testing to show the difference
    describe("successful call", () => {
      let iterator: Generator = createContactSaga(demoContact, []);
      it("should optimistically update", () => {
        expect(iterator.next().value).toMatchSnapshot("create-contact-optimistic-update");
      });
      it("should get the contacts", () => {
        expect(iterator.next().value).toEqual(call(createContact, demoContact));
      });
      it("should get replace the old contact with the new contact", () => {
        const contactWithId = {...demoContact, id: uuid};
        expect(iterator.next(contactWithId).value)
          .toMatchSnapshot("remove-old-snapshot");
      });
      it("should reset the new contact", () => {
        expect(iterator.next().value).toMatchSnapshot("create-contact-reset-after");
      });
      it("should close the create contact form", () => {
        expect(iterator.next().value).toMatchSnapshot("create-contact-close-form");
      });
      it("Should be done", () => {
        expect(iterator.next().done).toBeTruthy();
      });
    });
    describe("unsuccessful call", () => {
        it("should ask to retry", () => {
          let iterator: Generator = createContactSaga(demoContact, []);
          iterator.next();
          if (iterator.throw) {
            expect(iterator.throw(new Error("did not work")).value)
              .toEqual(call(confirm, "There was an error creating your new contact, retry?"));
            
          } else {
            throw new Error("something went wrong");
          }
        });

        describe("retry", () => {
          it("should retry on confirm", () => {
            let iterator: Generator; 
            iterator = createContactSaga(demoContact, []);
            iterator.next();
            if (iterator.throw) {
              iterator.throw(new Error("did not work"));
              expect(iterator.next(true).value)
              .toMatchSnapshot("create-contact-error-reset-list");
              expect(typeof iterator.next().value.next).toEqual("function");
            } else {
              throw new Error("something went wrong");
            }
             
          });
          it("should stop when the user does not retry", () => {
            let iterator: Generator; 
            iterator = createContactSaga(demoContact, []);
            iterator.next();
            if (iterator.throw) {
              iterator.throw(new Error("did not work"));
              expect(typeof iterator.next(false).value.next).toEqual("function");
            } else {
              throw new Error("something went wrong");
            }
          });
        });
    });
  });
});
