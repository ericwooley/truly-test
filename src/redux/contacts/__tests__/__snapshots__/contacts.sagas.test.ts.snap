// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-contact-close-form 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": false,
      "type": "DISPLAY_ADD_CONTACT_FORM",
    },
    "channel": null,
  },
}
`;

exports[`create-contact-error-reset-list 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Array [],
      "type": "SET_CONTACTS",
    },
    "channel": null,
  },
}
`;

exports[`create-contact-optimistic-update 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Array [
        Object {
          "context": "work",
          "name": "steve",
          "number": "1234567890",
        },
      ],
      "type": "SET_CONTACTS",
    },
    "channel": null,
  },
}
`;

exports[`create-contact-reset-after 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "RESET_NEW_CONTACT",
    },
    "channel": null,
  },
}
`;

exports[`remove-old-snapshot 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Array [
        Object {
          "context": "work",
          "id": "e02604b6-5e1d-402b-ae36-a966899fdcd9",
          "name": "steve",
          "number": "1234567890",
        },
      ],
      "type": "SET_CONTACTS",
    },
    "channel": null,
  },
}
`;
